src/mod/endpoints/mod_sofia/sofia.c

```diff
@@ -6962,7 +6962,27 @@ static void sofia_handle_sip_r_invite(switch_core_session_t *session, int status
 				}
 			}
 		}
-
+		
+		if(status == 200 && switch_channel_test_flag(channel, CF_RECOVERED) && !tech_pvt->recovered_call_route_fixed) {
```

Comment:
I would suggest to check the recovered call route fix via channel variable or a channel flag instead of a sofia variable. This will help investigate an issue where we can see these value in the generated cdr.

Reply:
Add a sofia config variable and a channel variable that can enable/disable this fix, in case of an issue we can disable it without rebuilding fs

Reply:
If possible check if the 200 OK has record-route on it so we use that instead of the recovered record routes.

Reply:
There is a problem with this suggestion:  The channel variable would persist across recovery, as it's part of the xml structure that is used to recover the call.  We could have 1 call survive more than 1 instance of recovery, and then on that 2nd or greater recovery die.  What I can do is simply make a chanvar that increments every time the dialog's route is fixed so that we know how many times it's occurred overall.

I can add a config variable and a chanvar override.

Normally a 200 OK to an in-dialog invite would not have a Record-Route header, but if it does, it should be ignored by the endpoints.  So there is no change that should be implemented here for this.

RFC 3261 section 12.2 Requests within a Dialog:

Requests within a dialog MAY contain Record-Route and Contact header
fields.  However, these requests do not cause the dialog's route set
to be modified, although they may modify the remote target URI.
Specifically, requests that are not target refresh requests do not
modify the dialog's remote target URI, and requests that are target
refresh requests do.  For dialogs that have been established with an
INVITE, the only target refresh request defined is re-INVITE (see
Section 14).  Other extensions may define different target refresh
requests for dialogs established in other ways.

	Note that an ACK is NOT a target refresh request.

Target refresh requests only update the dialog's remote target URI,
and not the route set formed from the Record-Route.  Updating the
latter would introduce severe backwards compatibility problems with
RFC 2543-compliant systems.
   
   
16.6 Request Forwarding (16 pertains to proxies)

If this request is already part of a dialog, the proxy SHOULD
insert a Record-Route header field value if it wishes to remain
on the path of future requests in the dialog.  In normal
endpoint operation as described in Section 12, these Record-
Route header field values will not have any effect on the route
sets used by the endpoints.

The proxy will remain on the path if it chooses to not insert a
Record-Route header field value into requests that are already
part of a dialog.  However, it would be removed from the path
when an endpoint that has failed reconstitutes the dialog.

A proxy MAY insert a Record-Route header field value into any
request.  If the request does not initiate a dialog, the
endpoints will ignore the value.  See Section 12 for details on
how endpoints use the Record-Route header field values to
construct Route header fields.